### PR TITLE
fix: Check L1DataFee in txpool promoteExecutables and demoteUnexecutables

### PR DESCRIPTION
**Pull request by vyzo**
## 1. Purpose or design rationale of this PR

When a user submits a transaction, [we check](https://github.com/scroll-tech/go-ethereum/blob/develop/core/tx_pool.go#L698) that the user's balance is enough to cover gas fee + L1 data fee in the txpool.

However, if L1GasPriceOracle is updated, it's possible that the user's balance is not sufficient anymore. Txpool does not repeat this check. As a result, such transactions are treated as pending (executable), and are forwarded to the worker for inclusion in the next block. Then we discard them in the worker with an insufficient balance error.

Tracing and processing these transactions is wasted effort. Instead, we should drop such transactions in promoteExecutables.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [X] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by vyzo** [comment]
Needs a test.

**Comment by vyzo** [comment]
I don't know what's up with this "CI/goimports-lint" check failing, it is acting up; imports have not changed.

**Comment by colinlyguo** [comment]
> I don't know what's up with this "CI/goimports-lint" check failing, it is acting up; imports have not changed.

it's a pending CI issue to be solved, can ignore it for now.

**Review by colinlyguo — DISMISSED** [review]
LGTM. why not adding `L1DataFee` check in `pool.demoteUnexecutables()`, too?

**Comment by vyzo** [comment]
probably a good idea, @Thegaram thoughts?

**Comment by vyzo** [comment]
Implemented in `demoteUnexecutables`

File: core/tx_list.go [hunk 1]
```diff
@@ -360,6 +360,31 @@ func (l *txList) Filter(costLimit *big.Int, gasLimit uint64) (types.Transactions
 	return removed, invalids
 }
 
+// FilterF removes all transactions from the list that satisfy a predicate.
+// Every removed transaction is returned for any
+// post-removal maintenance. Strict-mode invalidated transactions are also
+// returned.
```
Comment (colinlyguo) [inline core/tx_list.go]:
```suggestion
// FilterF removes all transactions from the list that satisfy a predicate.
// Every removed transaction is returned for any post-removal maintenance.
// Strict-mode invalidated transactions are also returned.
```

**Review by colinlyguo — DISMISSED** [review]
LGTM.

**Review by colinlyguo — DISMISSED** [review]

File: core/tx_list.go [hunk 2]
```diff
@@ -360,6 +360,30 @@ func (l *txList) Filter(costLimit *big.Int, gasLimit uint64) (types.Transactions
...
+// Every removed transaction is returned for any post-removal maintenance.
+// Strict-mode invalidated transactions are also returned.
+func (l *txList) FilterF(f func(tx *types.Transaction) bool) (types.Transactions, types.Transactions) {
+	removed := l.txs.Filter(f)
+
+	if len(removed) == 0 {
+		return nil, nil
+	}
+	var invalids types.Transactions
+	// If the list was strict, filter anything above the lowest nonce
```
Comment (0xmountaintop) [inline core/tx_list.go:392]:
```suggestion
	// If the list was strict, filter anything above the lowest removed nonce
```

**Review by 0xmountaintop — DISMISSED** [review]
LGTM just a nitpick

File: core/tx_list.go (hunk 2 above)
Comment (Thegaram) [reply core/tx_list.go:392]: Original wording is consistent with the copied geth code.

File: core/tx_pool.go [hunk 3]
```diff
@@ -1427,6 +1428,27 @@ func (pool *TxPool) promoteExecutables(accounts []common.Address) []*types.Trans
...
+	return func(tx *types.Transaction) bool {
+		if tx.Gas() > pool.currentMaxGas || tx.Cost().Cmp(costLimit) > 0 {
+			return true
+		}
+
+		if pool.chainconfig.Scroll.FeeVaultEnabled() {
+			// recheck L1 data fee, as the oracle price may have changed
+			l1DataFee, err := fees.CalculateL1DataFee(tx, pool.currentState)
+			if err != nil {
+				log.Trace("Failed to calculate L1 data fee", "err", err)
```
Comment (Thegaram) [inline core/tx_pool.go]:
```suggestion
				log.Error("Failed to calculate L1 data fee", "err", err, "tx", tx)
```

File: core/tx_list.go [hunk 4]
```diff
@@ -360,6 +360,30 @@ func (l *txList) Filter(costLimit *big.Int, gasLimit uint64) (types.Transactions
 	return removed, invalids
 }
 
+// FilterF removes all transactions from the list that satisfy a predicate.
+// Every removed transaction is returned for any post-removal maintenance.
+// Strict-mode invalidated transactions are also returned.
+func (l *txList) FilterF(f func(tx *types.Transaction) bool) (types.Transactions, types.Transactions) {
```
Comment (Thegaram) [inline core/tx_list.go]: What will be the effect of removing this part?

```go
// If all transactions are below the threshold, short circuit
if l.costcap.Cmp(costLimit) <= 0 && l.gascap <= gasLimit {
	return nil, nil
}
l.costcap = new(big.Int).Set(costLimit) // Lower the caps to the thresholds
l.gascap = gasLimit
```

**Review by Thegaram — DISMISSED** [review]

File: core/tx_list.go (hunk 4 above)
Comment (vyzo) [reply core/tx_list.go]: I dont know, so I am not inclined to remove it.

Comment (colinlyguo) [reply core/tx_list.go]: so that the code needs to be added to this function? the early return check.

Comment (Thegaram) [reply core/tx_list.go]: The decision is non-trivial. If we add the early return check, we won't check L1 data fee for those transactions. If we don't, then txpool will need to do much more work.

As for setting / not setting these two, I'm not sure about the impact:

```go
l.costcap = new(big.Int).Set(costLimit) // Lower the caps to the thresholds
l.gascap = gasLimit
```

Comment (colinlyguo) [reply core/tx_list.go]: yeah. I also noticed that the behavior of this heuristic changed (as that said, non-trivial) after considering the L1 data fee.
adding compatibility in this check requires changing the definition of `l.costcap` (adding L1 data fee):
```
l.txs.Put(tx)
if cost := tx.Cost(); l.costcap.Cmp(cost) < 0 {
	l.costcap = cost
}
```

Comment (vyzo) [reply core/tx_list.go]: alright, lets put it back.

I dont recall why i removed it, maybe it was by accident.

Comment (Thegaram) [reply core/tx_list.go]: > If we add the early return check, we won't check L1 data fee for those transactions

maybe this is why?

Comment (vyzo) [reply core/tx_list.go]: ah that makes sense

File: core/tx_pool.go [hunk 5]
```diff
@@ -1427,6 +1428,27 @@ func (pool *TxPool) promoteExecutables(accounts []common.Address) []*types.Trans
 	return promoted
 }
 
+func (pool *TxPool) executableTxFilter(addr common.Address) func(tx *types.Transaction) bool {
+	costLimit := pool.currentState.GetBalance(addr)
```
Comment (georgehao) [inline core/tx_pool.go]: Here maybe confront this case:

The caller logic 
```
f := pool.executableTxFilter(....)
// do some check, maybe need 20ms, like
Sleep(20ms)
// Then
f()
```
When f() begin to execute, the `costLimit` is `costLimit` before  20ms .

Why don't put `costLimit := pool.currentState.GetBalance(addr)` to `func(tx *types.Transaction) bool {`

File: core/tx_list.go (hunk 4 above)
Comment (colinlyguo) [reply core/tx_list.go]: We can consider adding the L1 data fee in `l.costcap`, thus the logic won't break anything.

File: core/tx_pool.go (hunk 5 above)
Comment (vyzo) [reply core/tx_pool.go]: thats the right place to put it i guess

Comment (georgehao) [reply core/tx_pool.go]: For current logic of scroll-tech/go-ethereum, put here is no issue. If caller call it like the above example, I think it's not right.

File: core/tx_list.go (hunk 4 above)
Comment (colinlyguo) [reply core/tx_list.go]: @vyzo Added a PR to implement the above logic, including compatibility: https://github.com/scroll-tech/go-ethereum/pull/681, and this PR needs more refactors.

bypassing the optimization may introduce some performance issues (although it's hard to know how exactly the performance would be influenced before deployment and running).

Comment (colinlyguo) [reply core/tx_list.go]: added in [8d383fb](https://github.com/scroll-tech/go-ethereum/pull/627/commits/8d383fb6b7a04e63cd34a1d08da0b9e5528cb601).

**Comment by colinlyguo** [comment]
CI failed in another unit test, can temporarily ignore it.

**Review by colinlyguo — DISMISSED** [review]

**Review by georgehao — DISMISSED** [review]

**Comment by colinlyguo** [comment]
Would we deploy this PR in one of the nodes to see if the executable transactions in tx pool can be pruned? cc @georgehao @Thegaram

**Review by Thegaram — DISMISSED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**